node_modules/col-activities/lib/api.js

```diff
@@ -0,0 +1,275 @@
+/**
+ * Copyright 2015 UC Berkeley (UCB) Licensed under the
+ * Educational Community License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may
+ * obtain a copy of the License at
+ *
+ *     http://opensource.org/licenses/ECL-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an "AS IS"
+ * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+var _ = require('lodash');
+var Joi = require('joi');
+
+var DB = require('col-core/lib/db');
+var log = require('col-core/lib/logger')('col-activities');
+
+var ActivitiesDefaults = require('./default');
+
+/* Activities */
+
+/**
+ * Create an activity that contributes points to the engagement index
+ *
+ * @param  {Course}           course                          The course to which the activity should be associated
+ * @param  {User}             user                            The user earning activity points for the activity
+ * @param  {String}           type                            The type of the activity. One of the types in `col-activities/lib/constants.js`
+ * @param  {Number}           objectId                        The id of the object on which the activity is taking place (e.g., the asset id, the comment id, etc.)
+ * @param  {String}           objectType                      The type of the object on which the activity is taking place. One of `CollabosphereConstants.ACTIVITY.OBJECT_TYPES`
+ * @param  {User}             [actor]                         The user performing the activity when different than the user earning activity points
+ * @param  {Function}         callback                        Standard callback function
+ * @param  {Object}           callback.err                    An error that occurred, if any
+ * @param  {Activity}         callback.activity               The created activity
+ */
+var createActivity = module.exports.createActivity = function(course, user, type, objectId, objectType, actor, callback) {
+  // Parameter validation
+  var validationSchema = Joi.object().keys({
+    'type': Joi.any().valid(_.pluck(ActivitiesDefaults, 'type')).required()
+  });
+
+  var validationResult = Joi.validate({
+    'type': type
+  }, validationSchema);
+
+  if (validationResult.error) {
+    return callback({'code': 400, 'msg': validationResult.error.details[0].message});
+  }
+
+  // Create the activity in the DB
+  var activity = {
+    'course_id': course.id,
+    'user_id': user.id,
+    'type': type,
+    'object_type': objectType
+  };
+  if (objectType === 'asset') {
+    activity.asset_id = objectId;
+  } else {
+    activity.object_id = objectId;
+  }
+  if (actor) {
+    activity.actor = actor.id;
+  }
+
+  DB.Activity.create(activity).complete(function(err, activity) {
```

Comment:
I've added a TODO as there are a number of places where transactions should be added